### PR TITLE
Plan viewer and approval for Copilot Chat

### PR DESCRIPTION
**Pull request by teresaqhoang**
### Motivation and Context

This PR enables Plan Viewer and Approval for Copilot chat. If SK plan is suggested on chatbot response generation, the plan will be returned to copilot chat for review before execution.

### Description

Flow

1. If ChatSkill suggests a Plan, that plan object will be returned to the client for user approval before it's run.
2. If the client receives a plan object back, it will provide a view only render of the plan in the chat window, and the user can approve or reject the plan.
3. If the user approves the plan, the User Intent determined from the first run (step 1) of the ChatSkill will be used, the plan will be invoked, and execution will continue as normal.
4. If the user rejects the plan, no action will be taken and chatbot response will be a constant string. 

Tech Details
WebApi
- Added new ProposedPlan class object (needs to be embedded so it can be parsed client side)

WebApp
- Added new Plan model objects

Screenshot 2023-05-09 at 7 06 31 PM

Screenshot 2023-05-09 at 7 09 10 PM

Screenshot 2023-05-09 at 7 09 18 PM

### CODE REVIEW

File: samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs
Comment (teresaqhoang) [inline samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs:340]: Do we want to have a pre-generated response like this or have the bot generate one? This way saves some time in response generation / roundtrips to the kernel

File: samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs
Comment (teresaqhoang) [inline samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs:421]: @TaoChenOSU

File: samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs
Comment (teresaqhoang) [inline samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs:63]: @TaoChenOSU this is my work around since the context variables aren't being populated as expected

**Comment by adrianwyatt** [comment]
Currently seeing the error `System.InvalidCastException: Unable to cast object of type 'CustomFunctionAsync' to type 'System.Func`2[Microsoft.SemanticKernel.Orchestration.SKContext,System.Threading.Tasks.Task`1[Microsoft.SemanticKernel.Orchestration.SKContext]]'.` using GitHub or Klarna plugins.

File: samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs
Comment (amsacha) [reply samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs:340]: Pre-generated will save on costing. Plus since the accept/cancel options are buttons, we arent getting more context from the user to drive a conversation. 
Perhaps reword it to "Sorry this plan didn't meet your goals. How shall I change this plan to better meet your requirements?"

File: samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs
Comment (lemillermicrosoft) [reply samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs:63]: Can you let me know more about the issues you are having, especially if you think they are an issue with `Plan`? Thanks!

**Review by lemillermicrosoft — APPROVED** [review]
Awesome stuff! I might reach out more about ideas, this will be nice to play with. Thanks!

File: samples/apps/copilot-chat-app/webapp/src/libs/models/Plan.ts
Comment (lemillermicrosoft) [inline samples/apps/copilot-chat-app/webapp/src/libs/models/Plan.ts:8]: I'm curious about this definition, I'd expect things like 'name' instead of 'function'.

Comment (lemillermicrosoft) [reply samples/apps/copilot-chat-app/webapp/src/libs/models/Plan.ts:8]: Ah I see the mapping in `sk-utilities`

Comment (lemillermicrosoft) [reply samples/apps/copilot-chat-app/webapp/src/libs/models/Plan.ts:8]: Something like

```
export type IPlan = {
    skill: string;
    function: string;
    description: string;
    stepInputs: IPlanInput[];
    steps: IPlan[];
};
```

might be better.

File: samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs
Comment (amsacha) [reply samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs:421]: Is there a reason we are trying to avoid storing the plan in semantic memory?

If a user cancels a plan they may then type up a prompt that asks the bot to change the plan in a certain step or how it approaches something over multiple steps. The LLM probably needs the previous plan to know how to modify it.

File: samples/apps/copilot-chat-app/webapp/src/libs/semantic-kernel/sk-utilities.ts
Comment (amsacha) [inline samples/apps/copilot-chat-app/webapp/src/libs/semantic-kernel/sk-utilities.ts:31]: Can we store the list of reserved keywords somewhere, and have a helper function do the check. I have a feeling as we interact with plans more, adding to context variables is going to happen in more places. It'll be easier to keep a track of it if its all in one place.

**Review by adrianwyatt — APPROVED** [review]

File: samples/apps/copilot-chat-app/webapp/src/libs/models/Plan.ts
Comment (teresaqhoang) [reply samples/apps/copilot-chat-app/webapp/src/libs/models/Plan.ts:8]: Sounds good -- I'll hold off on this and you can just merge your changes from your 📝 Refactor plan model and UI to support nested steps commit in after?

File: samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs
Comment (teresaqhoang) [reply samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs]: Resolved with Tao offline - will leave as is. It doesn't really matter if the plan is stored in semantic memory or not so we'll leave it to the bot to determine that

File: samples/apps/copilot-chat-app/webapp/src/libs/models/Plan.ts
Comment (lemillermicrosoft) [reply samples/apps/copilot-chat-app/webapp/src/libs/models/Plan.ts:8]: I'm not sure if/when any of that work will be going to `main` so I wouldn't count on it. I've already cherry-picked your UI work (and done this suggested change to support my scenarios) and will probably stay that way for the time being.

**Review by lemillermicrosoft — APPROVED** [review]

**Review by adrianwyatt — APPROVED** [review]

File: samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs
Comment (teresaqhoang) [reply samples/apps/copilot-chat-app/webapi/Skills/ChatSkill.cs:63]: The issue we were seeing is that the context variables aren't being propagated down the prompt chain as expected. 

It's not specifically a planner issue, but when I tried setting a new context variable 'proposedPlan' in AcquireExternalInformation method, it's not returned in the chatContext response in the ChatAsync method. 

Another observation to back this is the token limits aren't being passed down, i.e.,
tokenLimit is set in ExtractUserMemoriesAsync, so it should be lower or equal to that value in following methods, but it's not. 

One run yielded:
- tokenLimit = 2829 at the end of ExtractUserMemoriesAsync
- then tokenLimit = 2846 in AcquireExternalInformationAsync (which is called after)

**Pull request merged**
